video/out/vo_gpu_next.c

```diff
@@ -1509,112 +1513,84 @@ static void wait_events(struct vo *vo, int64_t until_time_ns)
     }
 }
 
-static char *get_cache_file(struct priv *p, char *type)
+#if PL_API_VER < 342
+static inline void xor_hash(void *hash, pl_cache_obj obj)
 {
-    char *file = NULL;
-    char *dir = NULL;
-    struct gl_video_opts *opts = p->opts_cache->opts;
+    *((uint64_t *) hash) ^= obj.key;
+}
 
-    if (strcmp(type, "shader") == 0) {
-        if (!opts->shader_cache)
-            goto done;
-        dir = opts->shader_cache_dir;
-    } else if (strcmp(type, "icc") == 0) {
-        if (!opts->icc_opts->cache)
-            goto done;
-        dir = opts->icc_opts->cache_dir;
-    } else {
-        goto done;
-    }
+static inline uint64_t pl_cache_signature(pl_cache cache)
+{
+    uint64_t hash = 0;
+    pl_cache_iterate(cache, xor_hash, &hash);
+    return hash;
+}
+#endif
+
+static void cache_init(struct vo *vo, struct cache *cache, size_t max_size,
+                       const char *dir_opt)
+{
+    struct priv *p = vo->priv;
+    const char *name = cache == &p->shader_cache ? "shader.cache" : "icc.cache";
 
-    if (dir && dir[0]) {
-        dir = mp_get_user_path(NULL, p->global, dir);
+    char *dir;
+    if (dir_opt && dir_opt[0]) {
+        dir = mp_get_user_path(vo, p->global, dir_opt);
     } else {
-        dir = mp_find_user_file(NULL, p->global, "cache", "");
-    }
-    if (dir && dir[0]) {
-        file = mp_path_join(NULL, dir, "libplacebo.cache");
-        mp_mkdirp(dir);
+        dir = mp_find_user_file(vo, p->global, "cache", "");
     }
-    talloc_free(dir);
-done:
-    return file;
-}
+    if (!dir || !dir[0])
+        return;
 
-static void load_cache_files(struct priv *p)
-{
-    char *icc_cache = get_cache_file(p, "icc");
-    char *shader_cache = get_cache_file(p, "shader");
-    bool same_cache = false;
-    if (icc_cache && shader_cache)
-        same_cache = strcmp(icc_cache, shader_cache) == 0;
-    if (shader_cache) {
-        FILE *cache = fopen(shader_cache, "rb");
-        if (cache) {
-            int ret = pl_cache_load_file(p->shader_cache, cache);
-            if (same_cache)
-                pl_cache_load_file(p->icc_cache, cache);
-            fclose(cache);
-            if (ret < 0)
-                MP_WARN(p, "Failed loading cache from %s\n", shader_cache);
-        }
-        talloc_free(shader_cache);
-    }
-    if (icc_cache && !same_cache) {
-        FILE *cache = fopen(icc_cache, "rb");
-        if (cache) {
-            int ret = pl_cache_load_file(p->icc_cache, cache);
-            fclose(cache);
-            if (ret < 0)
-                MP_WARN(p, "Failed loading cache from %s\n", icc_cache);
-        }
+    mp_mkdirp(dir);
+    cache->path = mp_path_join(NULL, dir, name);
```

Comment:
you forgot to parent this to `vo` (right now it's leaked)

Reply:
Decided to just parent this then, and free `dir` as usual (since it's only temporary)